### PR TITLE
Fix the heka grains for the aggregator/remote_collector

### PR DESCRIPTION
**Pull request by SwannCroiset**
Previously, the heka salt grains of the node running aggregator/remote_collector
get all the metric_collector alarms from all nodes (/etc/salt/grains.d/heka).
The resulting mines data is then wrong for the monitoring node, while that
situtation fortunately has no impact regarding metric_collector alarm
configurations, the Nagios service leverging mine data get a wrong list of
alarms for the monitoring node.

This patch fixes the issue with minimal changes but it appears that the logic
behind _service.sls state is not optimal and become hard to understand.
This state is executed several times with different contexts for every heka
'server' types and is not indempotent, indeed the /etc/salt/grains.d/heka file
content is different between 'local' servers (metric|log)_collector and
'remote' servers remote_collector|aggregator.

### CODE REVIEW

**Comment by SwannCroiset** [comment]
a simpler alternative to https://github.com/tcpcloud/salt-formula-heka/pull/59

File: heka/_service.sls [hunk 1]
```diff
@@ -119,8 +119,6 @@ heka_{{ service_name }}_service:
 
 {# Loading the other services' support metadata for local roles #}
 
-{%- if service_name in ['log_collector', 'metric_collector'] %}
-
 {%- for service_name, service in pillar.iteritems() %}
 {%- if service.get('_support', {}).get('heka', {}).get('enabled', False) %}
```
Comment (elemoine) [inline heka/_service.sls:123]: Now that we get all the services' support metadata for the remote_collector and the aggregator as well, you can remove the code section below where we read the Heka support metadata for the remote_collector and the aggregator.

File: heka/_service.sls [hunk 2]
```diff
@@ -145,9 +141,13 @@ heka_{{ service_name }}_service:
 {# Load the other services' support metadata from salt-mine #}
 
 {%- for node_name, node_grains in salt['mine.get']('*', 'grains.items').iteritems() %}
-{%- if node_grains.heka is defined %}
-{%- set service_grains = salt['grains.filter_by']({'default': service_grains}, merge=node_grains.heka) %}
-{%- endif %}
+{%- if node_grains.heka is defined and node_grains.heka is mapping %}
+{% for service, data in node_grains.heka.items() %}
```
Comment (elemoine) [inline heka/_service.sls:139]: Don't you want to use `{%-` here (with a dash)?

File: heka/_service.sls [hunk 3]
```diff
@@ -145,9 +141,13 @@ heka_{{ service_name }}_service:
 {# Load the other services' support metadata from salt-mine #}
 
 {%- for node_name, node_grains in salt['mine.get']('*', 'grains.items').iteritems() %}
-{%- if node_grains.heka is defined %}
-{%- set service_grains = salt['grains.filter_by']({'default': service_grains}, merge=node_grains.heka) %}
-{%- endif %}
+{%- if node_grains.heka is defined and node_grains.heka is mapping %}
```
Comment (elemoine) [inline heka/_service.sls]: Minor, but I would not bother checking that `node_grains.heka` is a mapping.

File: heka/_service.sls [hunk 4]
```diff
@@ -145,9 +141,13 @@ heka_{{ service_name }}_service:
...
+{% for service, data in node_grains.heka.items() %}
+  {%- if service in ['remote_collector', 'aggregator'] %}
```
Comment (elemoine) [inline heka/_service.sls]: Using a tuple rather than a list makes more sense to me. Again, this is minor.

**Review by elemoine — CHANGES_REQUESTED** [review]
I provided a few comments.

I understand that this patch fixes the problem. So I approve it. But the problem is that this relies on the `heka.remote_collector` and `heka.aggregator` states being applied after `heka.log_collector` and `heka.metric_collector`.

File: heka/_service.sls (hunk 1 above)
Comment (SwannCroiset) [reply heka/_service.sls:123]: right, done

File: heka/_service.sls (hunk 2 above)
Comment (SwannCroiset) [reply heka/_service.sls:139]: we don't really care about sls template rendering .. it is not a files I mean

Comment (elemoine) [reply heka/_service.sls:139]: Oh you're right.

File: heka/_service.sls (hunk 4 above)
Comment (SwannCroiset) [reply heka/_service.sls]: This is what we have so far .. a list but you are right

File: heka/_service.sls (hunk 3 above)
Comment (SwannCroiset) [reply heka/_service.sls]: why not ? I prefer check this is a dict() befor iterate over it.

Comment (elemoine) [reply heka/_service.sls]: I'd just rather fail if heka is not a mapping rather than silently skipping it.

**Review by elemoine — APPROVED** [review]
Let's merge this and see.

**Comment by SwannCroiset** [comment]
hold down:
```
---
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 10, in top-level template code
  File "/var/cache/salt/minion/files/base/heka/_service.sls", line 141, in top-level template code
    {%- set service_grains = salt['grains.filter_by']({'default': service_grains[service]}, merge=data) %}
  File "/usr/lib/python2.7/dist-packages/salt/modules/grains.py", line 546, in filter_by
    salt.utils.dictupdate.update(ret, copy.deepcopy(merge))
  File "/usr/lib/python2.7/dist-packages/salt/utils/dictupdate.py", line 35, in update
    raise TypeError('Cannot update using non-dict types in dictupdate.update()')
TypeError: Cannot update using non-dict types in dictupdate.update()


[CRITICAL] Rendering SLS 'base:heka.remote_collector' failed: Jinja error: Cannot update using non-dict types in dictupdate.update()
/var/cache/salt/minion/files/base/heka/_service.sls(141):
---
[...]

{%- for node_name, node_grains in salt['mine.get']('*', 'grains.items').iteritems() %}
{%- if node_grains.heka is defined %}
{% for service, data in node_grains.heka.items() %}
  {%- if service in ('remote_collector', 'aggregator') %}
    {%- set service_grains = salt['grains.filter_by']({'default': service_grains[service]}, merge=data) %}    <======================
  {%- endif %}
{% endfor %}
{% endif %}
{%- endfor %}

```

**Pull request merged**